### PR TITLE
Migration, CRUD, controllers, converter

### PR DESCRIPTION
**Pull request by Oleksandr-Eleks**

### CODE REVIEW

File: api/src/main/java/com/eleks/academy/pharmagator/scheduler/Scheduler.java [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
     private final List<DataProvider> dataProviderList;
 
-    @Scheduled(fixedDelay = 1, timeUnit = TimeUnit.MINUTES)
+    //@Scheduled(fixedDelay = 1, timeUnit = TimeUnit.MINUTES)
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/scheduler/Scheduler.java:23]: Revert this change

File: api/src/main/java/com/eleks/academy/pharmagator/services/MedicineServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                .map(source -> {
+                    Medicine m = modelMapper.map(medicineDto, Medicine.class);
+                    m.setId(id);
+                    medicineRepository.save(m);
+                    return m;
+                });
+    }
+
+    @Override
+    public void deleteById(Long id) {
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/services/MedicineServiceImpl.java:48]: Stay consistent with naming

File: api/src/main/java/com/eleks/academy/pharmagator/services/MedicineServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    public Medicine save(MedicineDto medicineDto) {
+        Medicine medicine = modelMapper.map(medicineDto, Medicine.class);
+        return medicineRepository.save(medicine);
+    }
+
+    @Override
+    public Optional<Medicine> update(Long id, MedicineDto medicineDto) {
+        return medicineRepository.findById(id)
+                .map(source -> {
+                    Medicine m = modelMapper.map(medicineDto, Medicine.class);
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/services/MedicineServiceImpl.java:40]: Please change 'm' to more meaningful

File: api/src/main/java/com/eleks/academy/pharmagator/services/MedicineServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                    m.setId(id);
+                    medicineRepository.save(m);
+                    return m;
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/services/MedicineServiceImpl.java:43]: Please return method result

File: api/src/main/java/com/eleks/academy/pharmagator/services/PharmacyServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public Optional<Pharmacy> findById(Long id) {
+        return pharmacyRepository.findById(id);
+    }
+
+    @Override
+    public Pharmacy save(PharmacyDto pharmacyDto) {
+        Pharmacy p = modelMapper.map(pharmacyDto, Pharmacy.class);
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/services/PharmacyServiceImpl.java:32]: Please change var name

File: api/src/main/java/com/eleks/academy/pharmagator/services/PharmacyServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public Optional<Pharmacy> update(Long id, PharmacyDto pharmacyDto) {
+        return pharmacyRepository.findById(id)
+                .map(ph ->{
+                    Pharmacy pharmacy = modelMapper.map(pharmacyDto, Pharmacy.class);
+                    pharmacy.setId(id);
+                    pharmacyRepository.save(pharmacy);
+                    return pharmacy;
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/services/PharmacyServiceImpl.java:43]: Please return result of save (update) operation

File: api/src/main/java/com/eleks/academy/pharmagator/services/PriceServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public Optional<Price> findById(Long pharmacyId, Long medicineId) {
+        PriceId priceId = new PriceId(pharmacyId, medicineId);
+        return this.priceRepository.findById(priceId);
+    }
+
+    @Override
+    public Price save(PriceDto priceDto) {
+        Price p = modelMapper.map(priceDto, Price.class);
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/services/PriceServiceImpl.java:39]: Please rename

File: api/src/main/java/com/eleks/academy/pharmagator/services/PriceServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    public Optional<Price> update(Long pharmacyId, Long medicineId, PriceDto priceDto) {
+        PriceId priceId = new PriceId(pharmacyId, medicineId);
+
+        return this.priceRepository.findById(priceId)
+                .map(pr -> {
+                    Price p = modelMapper.map(priceDto, Price.class);
+                    p.setPharmacyId(pharmacyId);
+                    p.setMedicineId(medicineId);
+                    priceRepository.save(p);
+                    return p;
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/services/PriceServiceImpl.java:54]: Please return result of update

File: api/src/main/java/com/eleks/academy/pharmagator/services/PriceServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        return priceRepository.save(p);
+    }
+
+    @Override
+    public Optional<Price> update(Long pharmacyId, Long medicineId, PriceDto priceDto) {
+        PriceId priceId = new PriceId(pharmacyId, medicineId);
+
+        return this.priceRepository.findById(priceId)
+                .map(pr -> {
+                    Price p = modelMapper.map(priceDto, Price.class);
```
Comment (Oleksandr-Eleks) [inline api/src/main/java/com/eleks/academy/pharmagator/services/PriceServiceImpl.java:50]: Rename? =)

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/MedicineController.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    public List<Medicine> getAll() {
+        return this.medicineService.findAll();
+    }
+
+    @GetMapping("/{id:[\\d]+}")
+    public ResponseEntity<Medicine> getById(@PathVariable Long id) {
+        return this.medicineService.findById(id)
+                .map(ResponseEntity::ok)
+                .orElseGet(() -> ResponseEntity.notFound().build());
+
```
[the remark this answers is not in the document]
Comment (Oleksandr-Eleks) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/MedicineController.java]: Outdated

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/MedicineController.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @PostMapping
+    public Medicine create(@Valid @RequestBody MedicineDto medicineDto) {
+        return this.medicineService.save(medicineDto);
+
```
[the remark this answers is not in the document]
Comment (ChernoVl) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/MedicineController.java]: done

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/PharmacyController.java [hunk 12]
```diff
@@ -1,24 +1,56 @@
...
+    public List<Pharmacy> getAll() {
+        return this.pharmacyService.findAll();
+    }
+
+    @GetMapping("/{id:[\\d]+}")
+    public ResponseEntity<Pharmacy> getById(@PathVariable Long id) {
+        return this.pharmacyService.findById(id)
+                .map(ResponseEntity::ok)
+                .orElseGet(() -> ResponseEntity.notFound().build());
+
```
[the remark this answers is not in the document]
Comment (ChernoVl) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/PharmacyController.java]: done

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/MedicineController.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    @PutMapping("/{id:[\\d]+}")
+    public ResponseEntity<Medicine> update(
+            @PathVariable Long id,
+            @Valid @RequestBody MedicineDto medicineDto) {
+
+        return this.medicineService.update(id, medicineDto)
+                .map(ResponseEntity::ok)
+                .orElseGet(() -> ResponseEntity.notFound().build());
+
```
[the remark this answers is not in the document]
Comment (ChernoVl) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/MedicineController.java]: done

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/PharmacyController.java [hunk 14]
```diff
@@ -1,24 +1,56 @@
...
+    }
+
+    @PostMapping
+    public ResponseEntity<Pharmacy> create(
+            @Valid @RequestBody PharmacyDto pharmacyDto) {
```
[the remark this answers is not in the document]
Comment (ChernoVl) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/PharmacyController.java]: okay 👌

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/PharmacyController.java [hunk 15]
```diff
@@ -1,24 +1,56 @@
...
+
+    @PutMapping("/{id:[\\d]+}")
+    public ResponseEntity<Pharmacy> update(
+            @PathVariable Long id,
+            @Valid @RequestBody PharmacyDto pharmacyDto) {
+
+        return this.pharmacyService.update(id, pharmacyDto)
+                .map(ResponseEntity::ok)
+                .orElseGet(() -> ResponseEntity.notFound().build());
+
```
[the remark this answers is not in the document]
Comment (ChernoVl) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/PharmacyController.java]: okay

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/PriceController.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+        return this.priceService.findById(pharmacyId, medicineId)
+                .map(ResponseEntity::ok)
+                .orElseGet(() -> ResponseEntity.notFound().build());
+    }
+
+    @PostMapping
+    public Price create(@Valid @RequestBody PriceDto priceDto) {
+        return this.priceService.save(priceDto);
+
```
[the remark this answers is not in the document]
Comment (ChernoVl) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/PriceController.java]: okay

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/PriceController.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+public class PriceController {
+
+    private final PriceService priceService;
+
+    @GetMapping
+    public List<Price> getAll() {
+        return this.priceService.findAll();
+    }
+
+    @GetMapping("/{pharmacyId:[\\d]+}/{medicineId:[\\d]+}")
```
[the remark this answers is not in the document]
Comment (ChernoVl) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/PriceController.java]: Thanks a lot. Done.

File: api/src/main/java/com/eleks/academy/pharmagator/controllers/PriceController.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+                .orElseGet(() -> ResponseEntity.notFound().build());
+    }
+
+    @PostMapping
+    public Price create(@Valid @RequestBody PriceDto priceDto) {
+        return this.priceService.save(priceDto);
+
+    }
+
+    @PostMapping("/{pharmacyId:[\\d]+}/{medicineId:[\\d]+}")
```
[the remark this answers is not in the document]
Comment (ChernoVl) [reply api/src/main/java/com/eleks/academy/pharmagator/controllers/PriceController.java]: done

**Review by zyoneal — APPROVED** [review]
